### ISSUE TITLE
Vim9: crash when using "silent!" and getting member fails from try/catch

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7416]
**Describe the bug**

In Vim9 script, failing to get a dictionary member from a `try`/`catch` and a `silent!` might make Vim crash.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        var d: dict<any>
        d = {'': 0}
        def Func()
            try
                eval '' .. map([''], {_, v -> {}[v]})->join() .. d['']
            catch
            endtry
        enddef
        sil! Func()
    EOF
    )

Vim crashes.

**Expected behavior**

Vim doesn't crash.

**Environment**

 - Vim version: 8.2 Included patches: 1-2083
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(362)

**Additional context**

Regression introduced in 8.2.2082.

---

- [backtrace](https://github.com/vim/vim/files/5633158/backtrace.txt)
- [valgrind log](https://github.com/vim/vim/files/5633159/valgrind.log)
- [asan log](https://github.com/vim/vim/files/5633157/asan.log)

---

Here is another snippet which causes Vim to crash:
```vim
vim9script
var d: dict<any>
def Func()
    try
        eval map([], {_, v -> {}[v]})->join() .. d['']
    catch
    endtry
enddef
sil! Func()
```
This one is caused by another regression introduced in 8.2.1230.

- [backtrace](https://github.com/vim/vim/files/5633190/backtrace.txt)
- [valgrind log](https://github.com/vim/vim/files/5633191/valgrind.log)
- [asan log](https://github.com/vim/vim/files/5633189/asan.log)

**Comment by lacygoill** [comment]
These are not the same crash as the one fixed in 8.2.2083.  I can still reproduce both of them while on the latter.

**Comment by lacygoill** [comment]
8.2.2090 has fixed the first crash, but not the second one:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        var d: dict<any>
        def Func()
            try
                eval map([], {_, v -> {}[v]})->join() .. d['']
            catch
            endtry
        enddef
        sil! Func()
    EOF
    )
